src/main/java/org/generationcp/middleware/service/impl/study/generation/ExperimentDesignServiceImpl.java

```diff
@@ -0,0 +1,240 @@
+			.getByStandardVariableId(new DmsProject(environmentDatasetId), TermId.EXPERIMENT_DESIGN_FACTOR.getId());
+		if (projectProp != null && projectProp.getValue() != null && NumberUtils.isDigits(projectProp.getValue())) {
+			return Optional.of(Integer.valueOf(projectProp.getValue()));
+		}
+		return Optional.absent();
+	}
+
+	private void saveVariables(final int studyId, final List<MeasurementVariable> variables, final Integer plotDatasetId, final List<Geolocation> geolocations) {
+		int plotDatasetNextRank = this.daoFactory.getProjectPropertyDAO().getNextRank(plotDatasetId);
+		final List<Integer> plotVariableIds = this.daoFactory.getProjectPropertyDAO().getVariableIdsForDataset(plotDatasetId);
+
+		final Integer envDatasetId = this.getEnvironmentDatasetId(studyId);
+		int envDatasetNextRank = this.daoFactory.getProjectPropertyDAO().getNextRank(envDatasetId);
+		final List<Integer> envVariableIds = this.daoFactory.getProjectPropertyDAO().getVariableIdsForDataset(envDatasetId);
+		// Save project variables in environment and plot datasets
+		for (final MeasurementVariable variable : variables) {
+			final int variableId = variable.getTermId();
+			final VariableType variableType = variable.getVariableType();
+			final boolean isEnvironmentVariable = VariableType.ENVIRONMENT_DETAIL.equals(variableType);
+			if (!this.variableExists(variableId, isEnvironmentVariable, envVariableIds, plotVariableIds)) {
+				Integer projectId = plotDatasetId;
+				Integer rank = 1;
+				Integer variableTypeId = variableType.getId();
+				if (isEnvironmentVariable) {
+					projectId = envDatasetId;
+					rank = envDatasetNextRank++;
+
+				} else {
+					// FIXME: Current bug that saves treatment factor as projectprop with type_id = 1100. Should be 1809
+					if (VariableType.TREATMENT_FACTOR.equals(variableType)) {
+						variableTypeId = TermId.MULTIFACTORIAL_INFO.getId();
+
+						// Save a record for same variable with variable type = EXPERIMENT_DESIGN
+						final DmsProject project = new DmsProject();
+						project.setProjectId(projectId);
+						final ProjectProperty property =
+							new ProjectProperty(project, VariableType.EXPERIMENTAL_DESIGN.getId(), "", plotDatasetNextRank++, variableId, variable.getAlias());
+						this.daoFactory.getProjectPropertyDAO().save(property);
+					}
+					rank = plotDatasetNextRank++;
+				}
+
+				final DmsProject project = new DmsProject();
+				project.setProjectId(projectId);
+				final String value = variable.getValue();
+				final ProjectProperty property =
+					new ProjectProperty(project, variableTypeId, value, rank, variableId, variable.getAlias());
+				this.daoFactory.getProjectPropertyDAO().save(property);
+
+				// FIXME Undo this duplicate saving in nd_geolocationprop (logged as part of IBP-3150)
+				if (isEnvironmentVariable) {
+					for (final Geolocation geolocation : geolocations) {
+						final GeolocationProperty geolocationProperty = new GeolocationProperty();
```

Comment:
does it helps to have a GeolocationProperty constructor passing these parameters? I think it will be more legible.